Exit status is set to 0 on failure instead of 1

When running tests with nose and the flaky plugin, the exit status was not being set to 1 when the only tests that failed were ones that run with the flaky decorator and had max tries greater than 1.  If a test that was not decorated with `@flaky` also failed, then the exit status was properly set to 1.  

For example, here is a test that always fails, run with nose and without flaky.  The exit status is `1`.
```
$ nosetests test_flaky.py:TestFlaky
F
-----------------------------------------------------------------------------
1) FAIL: test_flaky_exit (test_flaky.TestFlaky)

   Traceback (most recent call last):
    test_flaky.py line 7 in test_flaky_exit
      self.assertTrue(False)
   AssertionError: False is not true


-----------------------------------------------------------------------------
1 test run in 0.0 seconds. 
1 FAILED (0 tests passed)
$ echo $?
1
```

... And here is the same test, run with nose and using flaky with the default 2 max tries.  It correctly runs the test twice and prints the correct results, but then exits with `0` status.
```
$ nosetests test_flaky.py:TestFlaky --with-flaky
-F
-----------------------------------------------------------------------------
1) FAIL: test_flaky_exit (test_flaky.TestFlaky)

   Traceback (most recent call last):
    test_flaky.py line 7 in test_flaky_exit
      self.assertTrue(False)
   AssertionError: False is not true


-----------------------------------------------------------------------------
2 tests run in 0.0 seconds. 
1 FAILED, 1 skipped (0 tests passed)
$ echo $?
0
```

The root cause is that the flaky plugin passes a newly created `TextTestResult` object to `_rerun_test` rather than the original object that nose created, leaving nose unaware of any rerun results.

--- a/flaky/_flaky_plugin.py
+++ b/flaky/_flaky_plugin.py
@@ -377,7 +377,15 @@ class _FlakyPlugin(object):
         :rtype:
             `bool`
         """
-        no_retry = flaky[FlakyNames.CURRENT_RUNS] >= flaky[FlakyNames.MAX_RUNS]
+        max_runs, current_runs, min_passes, current_passes = (
+            flaky[FlakyNames.MAX_RUNS],
+            flaky[FlakyNames.CURRENT_RUNS],
+            flaky[FlakyNames.MIN_PASSES],
+            flaky[FlakyNames.CURRENT_PASSES],
+        )
+        runs_left = max_runs - current_runs
+        passes_needed = min_passes - current_passes
+        no_retry = passes_needed > runs_left
         return no_retry and not cls._has_flaky_test_succeeded(flaky)
 
     @staticmethod

--- a/flaky/flaky_nose_plugin.py
+++ b/flaky/flaky_nose_plugin.py
@@ -53,6 +53,10 @@ class FlakyPlugin(_FlakyPlugin, Plugin):
     def handleError(self, test, err):
         """
         Baseclass override. Called when a test raises an exception.
+
+        If the test isn't going to be rerun again, then report the error
+        to the nose test result.
+
         :param test:
             The test that has raised an error
         :type test:
@@ -75,6 +79,10 @@ class FlakyPlugin(_FlakyPlugin, Plugin):
     def handleFailure(self, test, err):
         """
         Baseclass override. Called when a test fails.
+
+        If the test isn't going to be rerun again, then report the failure
+        to the nose test result.
+
         :param test:
             The test that has raised an error
         :type test:
@@ -130,6 +138,9 @@ class FlakyPlugin(_FlakyPlugin, Plugin):
     def prepareTestResult(self, result):
         """
         Baseclass override. Called right before the first test is run.
+
+        Stores the test result so that errors and failures can be reported
+        to the nose test result.
 
         :param result:
             The nose test result that needs to be informed of test failures.

--- a/flaky/flaky_nose_plugin.py
+++ b/flaky/flaky_nose_plugin.py
@@ -21,6 +21,7 @@ class FlakyPlugin(_FlakyPlugin, Plugin):
         super(FlakyPlugin, self).__init__()
         self._logger = logging.getLogger('nose.plugins.flaky')
         self._flaky_result = TextTestResult(self._stream, [], 0)
+        self._nose_result = None
         self._flaky_report = True
         self._force_flaky = False
         self._max_runs = None
@@ -66,7 +67,10 @@ class FlakyPlugin(_FlakyPlugin, Plugin):
             `bool`
         """
         # pylint:disable=invalid-name
-        return self._handle_test_error_or_failure(test, err)
+        want_error = self._handle_test_error_or_failure(test, err)
+        if not want_error:
+            self._nose_result.addError(test, err)
+        return want_error
 
     def handleFailure(self, test, err):
         """
@@ -85,7 +89,10 @@ class FlakyPlugin(_FlakyPlugin, Plugin):
             `bool`
         """
         # pylint:disable=invalid-name
-        return self._handle_test_error_or_failure(test, err)
+        want_failure = self._handle_test_error_or_failure(test, err)
+        if not want_failure:
+            self._nose_result.addFailure(test, err)
+        return want_failure
 
     def addSuccess(self, test):
         """
@@ -119,6 +126,18 @@ class FlakyPlugin(_FlakyPlugin, Plugin):
         """
         if self._flaky_report:
             self._add_flaky_report(stream)
+
+    def prepareTestResult(self, result):
+        """
+        Baseclass override. Called right before the first test is run.
+
+        :param result:
+            The nose test result that needs to be informed of test failures.
+        :type result:
+            :class:`nose.result.TextTestResult`
+        """
+        # pylint:disable=invalid-name
+        self._nose_result = result
 
     def prepareTestCase(self, test):
         """

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -1,4 +1,6 @@
+-rrequirements.txt
 coveralls
+genty
 mock
 nose
 pep8

--- a/test/test_example.py
+++ b/test/test_example.py
@@ -29,7 +29,9 @@ class ExampleTests(TestCase):
         """
         self._threshold += 1
         if self._threshold < 1:
-            raise Exception("Threshold is not high enough.")
+            raise Exception("Threshold is not high enough: {0} vs {1}.".format(
+                self._threshold, 1),
+            )
 
     @flaky(3, 2)
     def test_flaky_thing_that_succeeds_then_fails_then_succeeds(self):
@@ -64,7 +66,9 @@ class ExampleFlakyTests(TestCase):
         """
         self._threshold += 1
         if self._threshold < 1:
-            raise Exception("Threshold is not high enough.")
+            raise Exception("Threshold is not high enough: {0} vs {1}.".format(
+                self._threshold, 1),
+            )
 
 
 def test_function():

--- a/test/test_flaky_nose_plugin.py
+++ b/test/test_flaky_nose_plugin.py
@@ -23,6 +23,8 @@ class TestFlakyPlugin(TestCase):
                 string_io.return_value = self._mock_stream
                 flaky_result.return_value = self._mock_test_result
                 self._flaky_plugin = flaky_nose_plugin.FlakyPlugin()
+        self._mock_nose_result = Mock(flaky_nose_plugin.TextTestResult)
+        self._flaky_plugin.prepareTestResult(self._mock_nose_result)
         self._mock_test = MagicMock(name='flaky_plugin_test')
         self._mock_test_case = MagicMock(
             name='flaky_plugin_test_case',
@@ -284,6 +286,7 @@ class TestFlakyPlugin(TestCase):
             }
         )
         expected_test_case_calls = [call.address(), call.address()]
+        expected_result_calls = []
         if expected_plugin_handles_failure:
             expected_test_case_calls.append(call.run(self._mock_test_result))
             expected_stream_calls = [call.writelines([
@@ -300,6 +303,14 @@ class TestFlakyPlugin(TestCase):
                 '\n',
             ])]
         else:
+            if is_failure:
+                expected_result_calls.append(
+                    call.addFailure(self._mock_test_case, self._mock_error),
+                )
+            else:
+                expected_result_calls.append(
+                    call.addError(self._mock_test_case, self._mock_error),
+                )
             expected_stream_calls = [call.writelines([
                 self._mock_test_method_name,
                 ' failed; it passed {0} out of the required {1} times.'.format(
@@ -314,6 +325,10 @@ class TestFlakyPlugin(TestCase):
                 unicode_type(self._mock_error[2]),
                 '\n'
             ])]
+        self.assertEqual(
+            self._mock_nose_result.mock_calls,
+            expected_result_calls,
+        )
         self.assertEqual(
             self._mock_test_case.mock_calls,
             expected_test_case_calls,

--- a/test/test_flaky_plugin.py
+++ b/test/test_flaky_plugin.py
@@ -1,10 +1,13 @@
 # coding: utf-8
 
 from __future__ import unicode_literals
+from genty import genty, genty_dataset
 from flaky._flaky_plugin import _FlakyPlugin
+from flaky.names import FlakyNames
 from test.base_test_case import TestCase
 
 
+@genty
 class TestFlakyPlugin(TestCase):
     def setUp(self):
         super(TestFlakyPlugin, self).setUp()
@@ -19,4 +22,37 @@ class TestFlakyPlugin(TestCase):
             mock_method_name,
             (ValueError.__name__, mock_exception, ''),
             mock_message
+        )
+
+    @genty_dataset(
+        default_not_started=(2, 1, 0, 0, False),
+        default_one_failure=(2, 1, 1, 0, False),
+        default_one_success=(2, 1, 1, 1, False),
+        default_two_failures=(2, 1, 2, 0, True),
+        default_one_failure_one_success=(2, 1, 2, 1, False),
+        three_two_not_started=(3, 2, 0, 0, False),
+        three_two_one_failure=(3, 2, 1, 0, False),
+        three_two_one_success=(3, 2, 1, 1, False),
+        three_two_two_failures=(3, 2, 2, 0, True),
+        three_two_one_failure_one_success=(3, 2, 2, 1, False),
+        three_two_two_successes=(3, 2, 2, 2, False),
+    )
+    def test_flaky_plugin_identifies_failure(
+            self,
+            max_runs,
+            min_passes,
+            current_runs,
+            current_passes,
+            expect_fail,
+    ):
+        flaky = {
+            FlakyNames.CURRENT_PASSES: current_passes,
+            FlakyNames.CURRENT_RUNS: current_runs,
+            FlakyNames.MAX_RUNS: max_runs,
+            FlakyNames.MIN_PASSES: min_passes,
+        }
+        # pylint:disable=protected-access
+        self.assertEqual(
+            self._flaky_plugin._has_flaky_test_failed(flaky),
+            expect_fail,
         )

--- a/test/test_options_example.py
+++ b/test/test_options_example.py
@@ -15,13 +15,13 @@ class ExampleTests(TestCase):
 
     def test_something_flaky(self):
         """
-        Flaky will run this test 3 times.
-        It will fail twice and then succeed once.
+        Flaky will run this test twice.
+        It will fail once and then succeed once.
         This ensures that we mark tests as flaky even if they don't have a
         decorator when we use the command-line options.
         """
         self._threshold += 1
-        if self._threshold < 1:
+        if self._threshold < 0:
             raise Exception("Threshold is not high enough.")
 
     @flaky(3, 1)

